projects/go-lib/src/lib/components/go-table/go-table-selection.model.ts

```diff
@@ -0,0 +1,41 @@
+export interface RowSelectionEvent extends SelectionState {
+  /**
+   * The current row that was targeted for selection
+   */
+  currentRow: {
+    /**
+     * The entire data in the row
+     */
+    data: object;
+    /**
+     * Whether or not this row was selected (`true`) or deselected (`false`)
+     */
+    selected: boolean;
+  };
+}
+
+export interface SelectionState {
+  /**
+   * The current deselected rows; if `selectionMode` is `selection` then this will always be an empty array
+   */
+  deselectedRows?: any[];
+
+  /**
+   * Defines the current mode of selection. `selection` will add items to `selectedRows`
+   * and `deselection` will add items to `deselectedRows`.
+   *
+   * If in `deselection` mode, the UI toggle all checkboxes to true. Items instead will be added to `deselectedRows`
+   * because they're being removed from the selection.
+   */
+  selectionMode: SelectionMode;
+
+  /**
+   * The current selected rows; if `selectionMode` is `deselection` then this will always be an empty array
+   */
+  selectedRows?: any[];
+}
+
+export enum SelectionMode {
```

Comment:
Is an enum the best way we can think of to store this selection mode? It doesn't seem very descriptive if I were to receive `1` or a `0` on the backend.

If we do decide this is the most simple way of doing this, do we want "selecting" to be a `0|false` or a `1|true`?

What are our other potential options? Maybe:
1. Storing this as a string?
    ```ruby
    {
      selectionMode: 'selecting' | 'deselecting'
    }
    ```
2. separating those out into two pieces of data?
    ```ruby
    {
      selecting: true | false,
      deselecting: true | false
    }
    ```
3. Maybe flipping the enum so the positives match up?
    ```javascript
    expect(selectionMode.deselection).toBe(0);
    expect(selectionMode.selection).toBe(1);
    ```

Reply:
@AlexOverbeck interesting, didn't really think about that. There are a couple approaches we can go with while still utilizing the `enum` as well to consider:

1. You can "stringify" the `enum` to the value of the enum and could send that to the backend, ex:
```
export enum SelectionMode {
  selection,
  deselection
}

// this will show message 'selection' as string representation of enum member
alert(SelectionMode[SelectionMode.selection]);
```

2. We could set the values equal to a string, ex:
```
export enum SelectionMode {
  selection = 'selection',
  deselection = 'deselection'
}

// this will show message 'selection' as string representation of enum member
alert(SelectionMode.selection);
```

I think the `enum` makes the front end implementation simpler and provides less room for error. Thought's on these approaches?

Reply:
@grahamhency, I think option number 2 here seems like it makes the most sense to me. What do you think?